### PR TITLE
Prove gateway service

### LINKED ISSUE #302: Implement ProveGateway service
Prove gateway
- Generate the account proof.
- Generate raw transaction for proof account
- Call transaction executor.

### PR DESCRIPTION
**Pull request by 0xsarvesh**
fixes #302  

Added proveGateway service

### CODE REVIEW

File: src/m1_facilitator/repositories/MessageRepository.ts [hunk 1]
```diff
@@ -216,6 +217,39 @@ export default class MessageRepository extends Subject<Message> {
     return this.convertToMessage(messageModel);
   }
 
+  /**
+   * This method returns messages based on below criteria.
+   *   - Filter based  on gateway address.
```
Comment (deepesh-kn) [inline src/m1_facilitator/repositories/MessageRepository.ts]:
```suggestion
   *   - Filter based on gateway address.
```

File: src/m1_facilitator/services/ProveGatewayService.ts [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+  private async proveGateway(
+    blockHeight: BigNumber,
+    pendingMessageType: MessageType,
+    gateway: Gateway,
+  ): Promise<void> {
+    let sourceWeb3;
+    let targetWeb3;
+    let transactionExecutor;
+
+    console.log('pendingMessageType  ', pendingMessageType);
```
Comment (deepesh-kn) [inline src/m1_facilitator/services/ProveGatewayService.ts]: Remove `console.log`.

File: test/m1_facilitator/repositories/MessageRepository/getMessageForConfirmation.test.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+  from '../../../../src/m1_facilitator/repositories/Repositories';
+import MessageRepository
+  from '../../../../src/m1_facilitator/repositories/MessageRepository';
+import Message, {
+  MessageStatus,
+  MessageType,
+} from '../../../../src/m1_facilitator/models/Message';
+import assert from '../../../test_utils/assert';
+
+describe('MessageRepository:getMessageForConfirmation', () => {
```
Comment (deepesh-kn) [inline test/m1_facilitator/repositories/MessageRepository/getMessageForConfirmation.test.ts]:
```suggestion
describe('MessageRepository:getMessageForConfirmation', (): void => {
```

File: test/m1_facilitator/services/ProveGateway/update.test.ts [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+  from '../../../../src/m1_facilitator/lib/TransactionExecutor';
+import Anchor from '../../../../src/m1_facilitator/models/Anchor';
+import Gateway, { GatewayType } from '../../../../src/m1_facilitator/models/Gateway';
+import Message, {
+  MessageStatus,
+  MessageType,
+} from '../../../../src/m1_facilitator/models/Message';
+import SpyAssert from '../../../test_utils/SpyAssert';
+
+describe('ProveGateway::update', () => {
```
Comment (deepesh-kn) [inline test/m1_facilitator/services/ProveGateway/update.test.ts]:
```suggestion
describe('ProveGateway::update', (): void => {
```

File: test/m1_facilitator/services/ProveGateway/update.test.ts [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+  let originTransactionExecutor: TransactionExecutor;
+  let auxiliaryTransactionExecutor: TransactionExecutor;
+  let proveGatewayService: ProveGatewayService;
+  let outboxProofSpy: any;
+  let getERC20CogatewaySpy: any;
+  let proveGatewaySpy: any;
```
Comment (deepesh-kn) [inline test/m1_facilitator/services/ProveGateway/update.test.ts:46]: Disable linter for L41,42,43
`// eslint-disable-next-line @typescript-eslint/no-explicit-any`

File: test/m1_facilitator/services/ProveGateway/update.test.ts [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+    outboxProofSpy = sinon.replace(
+      ProofGenerator.prototype,
+      'getOutboxProof',
+      sinon.fake.resolves(proof),
+    );
+
+    const fakeERC20Cogateway = {
+      methods: {
+        proveGateway: () => {
```
Comment (deepesh-kn) [inline test/m1_facilitator/services/ProveGateway/update.test.ts:99]: `// eslint-disable-next-line @typescript-eslint/explicit-function-return-type`

**Review by deepesh-kn — CHANGES_REQUESTED** [review]
Looks good, just linting issues.

File: src/m1_facilitator/repositories/MessageRepository.ts (hunk 1 above)
Comment (0xsarvesh) [reply src/m1_facilitator/repositories/MessageRepository.ts]: Done

File: src/m1_facilitator/services/ProveGatewayService.ts (hunk 2 above)
Comment (0xsarvesh) [reply src/m1_facilitator/services/ProveGatewayService.ts]: Done

File: test/m1_facilitator/repositories/MessageRepository/getMessageForConfirmation.test.ts (hunk 3 above)
Comment (0xsarvesh) [reply test/m1_facilitator/repositories/MessageRepository/getMessageForConfirmation.test.ts]: Done

File: test/m1_facilitator/services/ProveGateway/update.test.ts (hunk 4 above)
Comment (0xsarvesh) [reply test/m1_facilitator/services/ProveGateway/update.test.ts]: Done

File: test/m1_facilitator/services/ProveGateway/update.test.ts (hunk 5 above)
Comment (0xsarvesh) [reply test/m1_facilitator/services/ProveGateway/update.test.ts:46]: Done

File: test/m1_facilitator/services/ProveGateway/update.test.ts (hunk 6 above)
Comment (0xsarvesh) [reply test/m1_facilitator/services/ProveGateway/update.test.ts:99]: Done

**Review by deepesh-kn — APPROVED** [review]

**Pull request merged**